### ISSUE TITLE
Leap Year Problem?!

### ISSUE THREAD

**Issue by whitedavidp** [thess/callattendant#25]
Suddenly, today (2/29/24), an incoming call (any incoming call) causes Call Attendant to crash and burn. Here is what I see:

Feb 29 08:59:09 callattendant[2731]: Approved LED blinking: 2 times

Feb 29 08:59:09 callattendant[2731]: Waiting for call...

Feb 29 08:59:09 callattendant[2731]: Running the Flask server

Feb 29 08:59:09 callattendant[2731]:  * Serving Flask app 'callattendant'

Feb 29 08:59:09 callattendant[2731]:  * Debug mode: off

Feb 29 08:59:35 callattendant[2731]: RING LED blinking: 10 times

Feb 29 08:59:36 callattendant[2731]: > Queueing call 206xxxxxxx for processing

Feb 29 08:59:36 callattendant[2731]: Incoming call from 206xxxxxxx

Feb 29 08:59:36 callattendant[2731]: > Checking next-call-permitted

Feb 29 08:59:36 callattendant[2731]: > Checking whitelist(s)

Feb 29 08:59:36 callattendant[2731]: Approved LED blinking: 10 times

Feb 29 08:59:36 callattendant[2731]: ValueError('day is out of range for month')

Feb 29 08:59:36 callattendant[2731]: ** Error running callattendant

Feb 29 08:59:36 callattendant[2731]: Shutting down...

Feb 29 08:59:36 callattendant[2731]: -> Stopping modem

Feb 29 08:59:36 callattendant[2731]: -> Closing modem serial port

Feb 29 08:59:36 callattendant[2731]: -> Stopping voice mail

Feb 29 08:59:36 callattendant[2731]: -> Releasing resources

Feb 29 08:59:36 callattendant[2731]: Shutdown finished

Feb 29 08:59:36 systemd[1]: callattendant.service: Main process exited, code=exited, status=1/FAILURE

Feb 29 08:59:36 systemd[1]: callattendant.service: Failed with result 'exit-code'.

The issue would seem to be whatever is causing "ValueError('day is out of range for month')".

**Comment by whitedavidp** [comment]
Changing to development and debug I get:

Feb 29 09:13:14 callattendant[3074]: Adding to caller queue:

Feb 29 09:13:14 callattendant[3074]: {'DATE': '0229',

Feb 29 09:13:14 callattendant[3074]:  'NAME': 'WIRELESS CALLER',

Feb 29 09:13:14 callattendant[3074]:  'NMBR': '206xxxxxxx',

Feb 29 09:13:14 callattendant[3074]:  'TIME': '0913'}

Feb 29 09:13:14 callattendant[3074]: Incoming call from 206xxxxxxx

Feb 29 09:13:14 callattendant[3074]: > Checking next-call-permitted

Feb 29 09:13:14 callattendant[3074]: > Checking whitelist(s)

Feb 29 09:13:14 callattendant[3074]: Approved LED blinking: 10 times

Feb 29 09:13:14 callattendant[3074]: ValueError('day is out of range for month')

Feb 29 09:13:14 callattendant[3074]: ** Error running callattendant

Feb 29 09:13:14 callattendant[3074]: Shutting down...

Feb 29 09:13:14 callattendant[3074]: -> Stopping modem

Feb 29 09:13:14 callattendant[3074]: -> Closing modem log file

Feb 29 09:13:14 callattendant[3074]: -> Closing modem serial port

Feb 29 09:13:14 callattendant[3074]: -> Stopping voice mail

Feb 29 09:13:14 callattendant[3074]: Message Event triggered

Feb 29 09:13:14 callattendant[3074]: -> Releasing resources

Feb 29 09:13:14 callattendant[3074]: Shutdown finished

Feb 29 09:13:14 systemd[1]: callattendant.service: Main process exited, code=exited, status=1/FAILURE

Feb 29 09:13:14 systemd[1]: callattendant.service: Failed with result 'exit-code'.

**Comment by roycewilliams** [comment]
Confirmed on my installs as well.
[2024-02-29 08:33:26.789280] ValueError('day is out of range for month')

Long shot, but I saw that tzdata was out of date, upgraded - no change.

**Comment by whitedavidp** [comment]
Thanks for the confirmation. The month/year is confirmed to match on my one, caller-id enabled handset. A quick grep turns up many datetime references and it is not clear to me which might be causing this. But I suspect it is very early-on in call processing. The call log in the database does not get updated.

I suppose one day of downtime every 4 years (not quite that frequently, I know) isn't so bad. But I suspect this can be fixed by someone more expert than I am.

**Comment by thess** [comment]
Well, something to do after lunch today ;)

I'll try to get it fixed before the next leap-year.

**Comment by whitedavidp** [comment]
As always, thanks. I have tried a bit to run it to ground without any real progress...

**Comment by thess** [comment]
Ha, ha, ha -- Phone company only reports the month and day and not the year. Python `datetime` functions default to 1900 if no year specified and there was no leap-year then (not divisible by 400) ! Stay tuned for fix.

**Comment by thess** [comment]
See fix in `calllogger.py` -- 652b33b

Will include in 1.7.1 release TBS

**Comment by whitedavidp** [comment]
That seems to fix it. Thanks so much!

Man- it was getting ugly today with all the junk calls and NO callattendant!

**Comment by roycewilliams** [comment]
To apply the fix now, this command-line snippet may help (you may need to adjust your python path):

```
cd ~/venv/lib/python3.9/site-packages/callattendant/screening/
wget https://github.com/thess/callattendant/commit/652b33b1eed76f77393bac2fecb4b31f3ea0715c.patch
cp -p calllogger.py calllogger.py-dist
patch calllogger.py 652b33b1eed76f77393bac2fecb4b31f3ea0715c.patch
```

**Comment by thess** [comment]
FYI - this code was added to the original repo in August 2020. Never encountered leap-year before today.

**Comment by thess** [comment]
Closed in release v1.7.3
